README.md
# flomod
[![The GPL License](https://img.shields.io/badge/license-GPL-yellow.svg?style=flat-square)](https://opensource.org/licenses/GPL-3.0)
[![Travis CI](https://img.shields.io/travis/Jacajack/flomod.svg?style=flat-square)](https://travis-ci.org/Jacajack/flomod)

[Flomod on launchpad...](https://launchpad.net/flomod)

`flomod` is a floppy disk image editor - it allows user to read or modify given sectors, tracks or bytes on floppy disk or its image.
It's relatively easy to use and can be pretty powerful tool when used along with other UNIX commands.

## Installing flomod
 - Add PPA to your system - `sudo add-apt-repository ppa:mrjjot/mrjjot`
 - Update software lists - `sudo apt-get update`
 - Install `flomod` package - `sudo apt-get install flomod`


## Usage
`flomod [OPTIONS] diskfile.bin`

Supported command line options:
 - `-1` / `--baseone` - Sector numbers begin with 1 (like in BIOS interrupt 13h)
 - `-s` / `--start` - Specify start point of read/write operation in C:H:S:B format
 - `-e` / `--end` - Specify end point of read/write operation in C:H:S:B format (overrides `-n` flag)
 - `-n` / `--num` - Specify length of read/write operation in C:H:S:B format
 - `-g` / `--geom` - Manually define disk geometry
 - `-t` / `--type` - Specify disk geometry preset
 - `-d` / `--dump` - Display written/read data along with its addresses in hex format
 - `-v` / `--verbose` - Be more verbose about what's going on
 - `-w` / `--write` - Write data from `stdin` to the image file
 - `-r` / `--read` - Read data from the image file

Disk geometry presets include:
 - `FLOPPY_3.5_740K` - 3.5 inch floppy, 740KB (equivalent of 80:2:9:512)
 - `FLOPPY_3.5_1.44M` - 3.5 inch floppy, 1.44MB (equivalent of 80:2:18:512)
 - `FLOPPY_3.5_2.88M` - 3.5 inch floppy, 2.88MB (equivalent of 80:2:36:512)
 - `FLOPPY_5.25_360K` - 5.25 inch floppy, 360KB (equivalent of 80:2:15:512)
 - `FLOPPY_5.25_1.2M` - 5.25 inch floppy, 1.2MB (equivalent of 40:2:9:512)

For example, if you want to read sector 3 from track 7 (side 2), you can use following command:<br>`flomod -s 7:1:3:0 -n 0:0:1:0 floppy.bin`.

## Building from source
 - Clone git repository - `git clone https://github.com/Jacajack/flomod.git && cd flomod`
 - Invoke makefile - `make all`
 - To install use `sudo make install`, and to uninstall - `sudo make uninstall`.


debian/changelog
flomod (0.6-3) trusty; urgency=low

  * Minor docs changes.

 -- Jacek Wieczorek <mrjjot@gmail.com>  Sat, 8 Jul 2017 17:13:00 +0200


flomod (0.6-2) trusty; urgency=low

  * Fix disk geometry checking bug.

 -- Jacek Wieczorek <mrjjot@gmail.com>  Wed, 28 Jun 2017 02:20:00 +0200

flomod (0.6-1) trusty; urgency=low

  * Fix some bugs.
  * Rename geometry flags.
  * Add some warnings.
  * Improve docs and manpages.

 -- Jacek Wieczorek <mrjjot@gmail.com>  Wed, 28 Jun 2017 00:13:00 +0200

flomod (0.1-1) trusty; urgency=low

  * Initial release.

 -- Jacek Wieczorek <mrjjot@gmail.com>  Tue, 27 Jun 2017 15:18:00 +0200


src/args.c
/*
	flomod - a floppy disk image editor
	Copyright (C) 2017 Jacek Wieczorek <mrjjot@gmail.com>
	This file is part of flomod.
	Flomod is free software: you can redistribute it and/or modify
	it under the terms of the GNU General Public License as published by
	the Free Software Foundation, either version 3 of the License, or
	(at your option) any later version.
	Flomod is distributed in the hope that it will be useful,
	but WITHOUT ANY WARRANTY; without even the implied warranty of
	MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
	GNU General Public License for more details.
	You should have received a copy of the GNU General Public License
	along with this program.  If not, see <http://www.gnu.org/licenses/>.
*/

#include <argp.h>
#include "flomod.h"

const char *argp_program_version = "flomod v0.6";
const char *argp_program_bug_address = "<mrjjot@gmail.com>";
char argp_doc[] = "flomod - A program to edit floppy disk.";
char argp_keydoc[] = "FILENAME";
struct argp_option argp_options[] =
{
	{"read",   'r', 0,          0, "Read data from disk"},
	{"write",  'w', 0,          0, "Write data to disk"},
	{"dump",   'd', 0,          0, "Dump read/written data in hex format"},
	{"verbose",'v', 0,          0, "Be verbose"},
	{"baseone",'1', 0,          0, "Use 1 based sector numeration (like int 13h)"},
	{"start",  's', "C:H:S:B",  0, "Set start point for disk operation"},
	{"end",    'e', "C:H:S:B",  0, "Set end point for disk operation"},
	{"num",    'n', "C:H:S:B",  0, "Set disk operation length"},
	{"geom",   'g', "C:H:S:B",  0, "Set disk geometry limits"},
	{"type",   't', "TYPE",     0, "Set disk type (see manpage for list)"},

	{0}
};

//Lovely callback for ARGP to parse command line options
error_t parse_opt( int key, char *arg, struct argp_state *state )
{
	struct flomod *conf = (struct flomod*) state->input;

	switch ( key )
	{
		case 'v':
			conf->flags |= FLOMOD_FLAG_VERBOSE;
			break;

		case 'r':
			conf->flags &= ~FLOMOD_FLAG_WRITE;
			break;

		case 'w':
    		conf->flags |= FLOMOD_FLAG_WRITE;
    		break;

		case 'd':
			conf->flags |= FLOMOD_FLAG_DUMP;
			break;

		case '1':
			conf->flags |= FLOMOD_FLAG_SECTOR_BASE1;
			break;

		case 's':
    		conf->start.str = arg;
    		break;

		case 'e':
			conf->end.str = arg;
			break;

		case 'n':
	    	conf->length.str = arg;
	    	break;

		case 'g':
			conf->geom.str = arg;
			break;

		case 't':
			conf->disktype = arg;
			break;

		case ARGP_KEY_ARG:
    		if ( state->arg_num >= 1 ) argp_usage( state );
			conf->diskfname = arg;
			break;

		case ARGP_KEY_END:
    		if ( state->arg_num < 1 ) argp_usage( state );
    		break;

    	default:
    		return ARGP_ERR_UNKNOWN;
    }
	return 0;
}


src/args.h
/*
	flomod - a floppy disk image editor
	Copyright (C) 2017 Jacek Wieczorek <mrjjot@gmail.com>
	This file is part of flomod.
	Flomod is free software: you can redistribute it and/or modify
	it under the terms of the GNU General Public License as published by
	the Free Software Foundation, either version 3 of the License, or
	(at your option) any later version.
	Flomod is distributed in the hope that it will be useful,
	but WITHOUT ANY WARRANTY; without even the implied warranty of
	MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
	GNU General Public License for more details.
	You should have received a copy of the GNU General Public License
	along with this program.  If not, see <http://www.gnu.org/licenses/>.
*/

#ifndef ARGS_H
#define ARGS_H
#include <argp.h>

extern const char *argp_program_version;
extern const char *argp_program_bug_address;
extern char argp_doc[];
extern char argp_keydoc[];
extern struct argp_option argp_options[];
extern error_t parse_opt( int key, char *arg, struct argp_state *state );

#endif


src/chsb.c
/*
	flomod - a floppy disk image editor
	Copyright (C) 2017 Jacek Wieczorek <mrjjot@gmail.com>
	This file is part of flomod.
	Flomod is free software: you can redistribute it and/or modify
	it under the terms of the GNU General Public License as published by
	the Free Software Foundation, either version 3 of the License, or
	(at your option) any later version.
	Flomod is distributed in the hope that it will be useful,
	but WITHOUT ANY WARRANTY; without even the implied warranty of
	MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
	GNU General Public License for more details.
	You should have received a copy of the GNU General Public License
	along with this program.  If not, see <http://www.gnu.org/licenses/>.
*/

#include <stdio.h>
#include <inttypes.h>
#include <stdlib.h>
#include <string.h>
#include "chsb.h"

void chsbopt( CHSB *desc, CHSB *lim )
{
	if ( desc == NULL || lim == NULL ) return;

	//Optimize byte number
	desc->s += desc->b / lim->b;
	desc->b = desc->b % lim->b;
	if ( desc->b < 0 ) desc->b = 0;

	//Optimize sector number
	if ( desc->flags & CHSB_FLAG_SECTOR_BASE1 )
	{
		//1-based sector numbers
		desc->h += ( desc->s - 1 ) / lim->s;
		desc->s = ( desc->s - 1 ) % lim->s + 1;
		if ( desc->s < 1 ) desc->s = 1;
	}
	else
	{
		//0-based sector numbers
		desc->h += desc->s / lim->s;
		desc->s = desc->s % lim->s;
		if ( desc->s < 0 ) desc->s = 0;
	}
	//Optimize head number
	desc->c += desc->h / lim->h;
	desc->h = desc->h % lim->h;
	if ( desc->h < 0 ) desc->h = 0;

	//Check cylinder number bounds
	if ( desc->c < 0 ) desc->c = 0;
}

//Sum two CHSB structures (optimization not included)
void chsbsum( CHSB *desc, CHSB *offset )
{
	if ( desc == NULL || offset == NULL ) return;
	desc->c += offset->c;
	desc->h += offset->h;
	desc->s += offset->s - ( desc->flags & CHSB_FLAG_SECTOR_BASE1 ? 1 : 0 );
	desc->b += offset->b;
}

//Check if the CHSB struct is empty
int chsbnull( CHSB *desc )
{
	if ( desc == NULL ) return 1;
	return 	desc->c == 0 && \
			desc->h == 0 && \
			( desc->s == 0 || ( desc->s == 1 && desc->flags & CHSB_FLAG_SECTOR_BASE1 ) ) && \
			desc->b == 0;
}

//Convert CHSB structure to offset in sectors or bytes
void chsb2lba( CHSB *desc, CHSB *lim )
{
	if ( desc == NULL || lim == NULL ) return;
	desc->offset = ( desc->c * lim->h + desc->h ) * lim->s + desc->s;
	if ( desc->flags & CHSB_FLAG_SECTOR_BASE1 ) desc->offset -= 1;
	desc->offset = desc->offset * lim->b + desc->b;
}

//Convert offset in sectors or bytes to CHSB structure values
void lba2chsb( CHSB *desc, CHSB *lim )
{
	if ( desc == NULL || lim == NULL ) return;
	desc->c = desc->h = 0;
	desc->s = desc->flags & CHSB_FLAG_SECTOR_BASE1 ? 1 : 0;
	desc->b = desc->offset;
	chsbopt( desc, lim );
}

//Convert colon-separated CHSB string to CHSB structure
int str2chsb( CHSB *desc )
{
	unsigned char ec = 1;
	const char *delim = ":";
	char *str = NULL;
	char *token = NULL;

	if ( desc == NULL ) return 1;
	desc->c = 0;
	desc->h = 0;
	desc->s = desc->flags & CHSB_FLAG_SECTOR_BASE1 ? 1 : 0;
	desc->b = 0;

	if ( desc->str == NULL ) return 1;
	str = strdup( desc->str );
	if ( str == NULL ) return 1;

	token = strtok( str, delim );
	if ( token != NULL && !sscanf( token, "%ld", &desc->c ) ) desc->c = 0;
	else ec = 0;

	token = strtok( NULL, delim );
	if ( token != NULL && !sscanf( token, "%ld", &desc->h ) ) desc->h = 0;
	else ec = 0;

	token = strtok( NULL, delim );
	if ( token != NULL && !sscanf( token, "%ld", &desc->s ) ) desc->s = 0;
	else ec = 0;

	token = strtok( NULL, delim );
	if ( token != NULL && !sscanf( token, "%ld", &desc->b ) ) desc->b = 0;
	else ec = 0;

	free( str );
	return ec;
}

//Generate new string representation for the CHSB structure
const char *chsb2str( CHSB *desc )
{
	const char *delim = ":";
	char str[64];

	if ( desc == NULL ) return NULL;

	snprintf( str, 63, "%ld%s%ld%s%ld%s%ld", desc->c, delim, desc->h, delim, desc->s, delim, desc->b );

	desc->str = strdup( str );
	return desc->str;
}


src/chsb.h
/*
	flomod - a floppy disk image editor
	Copyright (C) 2017 Jacek Wieczorek <mrjjot@gmail.com>
	This file is part of flomod.
	Flomod is free software: you can redistribute it and/or modify
	it under the terms of the GNU General Public License as published by
	the Free Software Foundation, either version 3 of the License, or
	(at your option) any later version.
	Flomod is distributed in the hope that it will be useful,
	but WITHOUT ANY WARRANTY; without even the implied warranty of
	MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
	GNU General Public License for more details.
	You should have received a copy of the GNU General Public License
	along with this program.  If not, see <http://www.gnu.org/licenses/>.
*/

#ifndef CHSB_H
#define CHSB_H
#include <inttypes.h>

#define CHSB_FLAG_SECTOR_BASE1 1

//CHSB structure
typedef struct
{
	long c;
	long h;
	long s;
	long b;
	long offset;
	const char *str;
	unsigned int flags;
} CHSB;

extern void chsbopt( CHSB *desc, CHSB *lim );
extern void chsbsum( CHSB *desc, CHSB *val );
extern int chsbnull( CHSB *desc );
extern void chsb2lba( CHSB *desc, CHSB *lim );
extern void lba2chsb( CHSB *desc, CHSB *lim );
extern int str2chsb( CHSB *desc );
extern const char *chsb2str( CHSB *desc );

#endif


src/flomod.c
/*
	flomod - a floppy disk image editor
	Copyright (C) 2017 Jacek Wieczorek <mrjjot@gmail.com>
	This file is part of flomod.
	Flomod is free software: you can redistribute it and/or modify
	it under the terms of the GNU General Public License as published by
	the Free Software Foundation, either version 3 of the License, or
	(at your option) any later version.
	Flomod is distributed in the hope that it will be useful,
	but WITHOUT ANY WARRANTY; without even the implied warranty of
	MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
	GNU General Public License for more details.
	You should have received a copy of the GNU General Public License
	along with this program.  If not, see <http://www.gnu.org/licenses/>.
*/

#include <stdio.h>
#include <inttypes.h>
#include <string.h>
#include <stdlib.h>
#include <unistd.h>
#include <ctype.h>
#include <argp.h>
#include "chsb.h"
#include "disktypes.h"
#include "args.h"
#include "flomod.h"

//Main configuration structure (all flags etc.)
struct flomod flomod;

void dump( long start, long current, unsigned char b )
{
	#define DUMP_WIDTH 8
	static unsigned char buf[DUMP_WIDTH];
	CHSB pos = {0};
	int i;

	//If we are at line begining
	if ( ( current - start ) % DUMP_WIDTH == 0 )
	{
		if ( current == start )
		{
			//If we are at start - print lovely header
			printf( "| cyl  | head | sect | byte | ---\n" );
		}
		else printf( "\n" );

		//Convert LBA to CHSB
		pos.offset = current;
		if ( flomod.flags & FLOMOD_FLAG_SECTOR_BASE1 ) pos.flags |= CHSB_FLAG_SECTOR_BASE1;
		lba2chsb( &pos, &flomod.geom );

		//Print current position (CHSB)
		printf( "|%6ld|%6ld|%6ld|%6ld| ", pos.c, pos.h, pos.s, pos.b );
	}

	//Store current byte in buffer and print it as hex
	buf[( current - start ) % DUMP_WIDTH] = b;
	printf( "%02x ", b );

	//If we are at last byte in the line - dump buffer as characters
	if ( ( current - start ) % DUMP_WIDTH == DUMP_WIDTH - 1 )
	{
		printf( "| " );

		//Dump whole buffer, but skip whitespace characters
		for ( i = 0; i < DUMP_WIDTH; i++ )
			if ( !isspace( buf[i] ) )
			{
				printf( "%c", buf[i] );
			}
			else printf( "." );
	}
	#undef DUMP_WIDTH
}

int main( int argc, char **argv )
{
	int i, b;

	//Hardcoded executable name (I'm sorry)
	flomod.exename = "flomod";

	//Parse command line arguments
	static struct argp argp = {argp_options, parse_opt, argp_keydoc, argp_doc};
	argp_parse( &argp, argc, argv, 0, 0, &flomod );

	//Setup CHSB limits
	str2chsb( &flomod.geom );
	if ( chsbnull( &flomod.geom ) ) //Skip if set by user
	{
		//Default disk type if not specified
		if ( flomod.disktype == NULL ) flomod.flags |= FLOMOD_FLAG_DEFAULT_FLOPPY;
		else
		{
			//If disk type specified - find it
			for ( i = 0; i < DISKTYPE_COUNT; i++ )
			{
				if ( !strcmp( disktypes[i].str, flomod.disktype ) )
				{
					flomod.geom = disktypes[i];
					break;
				}
			}

			//Unknown disk type - quit
			if ( i == DISKTYPE_COUNT )
			{
				fprintf( stderr, "%s: unknown disk type\n", flomod.exename );
				exit( 1 );
			}
		}
	}

	//Default disk type
	if ( flomod.flags & FLOMOD_FLAG_DEFAULT_FLOPPY )
	{
		flomod.geom = disktypes[DISKTYPE_FLOPPY_35_144];
		if ( flomod.flags & FLOMOD_FLAG_VERBOSE ) fprintf( stderr, "%s: default disk type is 3.5\" floppy\n", flomod.exename );
	}

	//Check for invalid disk geometry
	if ( flomod.geom.c <= 0 || flomod.geom.h <= 0 || flomod.geom.s <= 0 || flomod.geom.b <= 0 )
	{
		fprintf( stderr, "%s: invalid disk geometry!\n", flomod.exename );
		exit( 1 );
	}

	//Base 1 sector numeration
	if ( flomod.flags & FLOMOD_FLAG_SECTOR_BASE1 )
	{
		flomod.start.flags |= CHSB_FLAG_SECTOR_BASE1;
		flomod.length.flags |= CHSB_FLAG_SECTOR_BASE1;
		flomod.end.flags |= CHSB_FLAG_SECTOR_BASE1;
		flomod.geom.flags |= CHSB_FLAG_SECTOR_BASE1;
	}

	//Get start descriptor
	str2chsb( &flomod.start );
	chsbopt( &flomod.start, &flomod.geom );
	chsb2lba( &flomod.start, &flomod.geom );
	chsb2str( &flomod.start );

	//Get length
	str2chsb( &flomod.length );
	chsbopt( &flomod.length, &flomod.geom );
	chsb2lba( &flomod.length, &flomod.geom );

	//No filename
	if ( flomod.diskfname == NULL )
	{
		if ( flomod.flags & FLOMOD_FLAG_VERBOSE ) fprintf( stderr, "%s: disk file not specified!\n", flomod.exename );
		exit( 1 );
	}

	//Open disk file
	if ( flomod.flags & FLOMOD_FLAG_WRITE )
	{
		//Attempt to open for writing
		if ( access( flomod.diskfname, F_OK ) != -1 )
			flomod.diskf = fopen( flomod.diskfname, "r+b" ); //If file exists
		else
		{
			if ( flomod.flags & FLOMOD_FLAG_VERBOSE ) fprintf( stderr, "%s: creating '%s'\n", flomod.exename, flomod.diskfname );
			flomod.diskf = fopen( flomod.diskfname, "wb" ); //If file doesn't exist
		}
	}
	else
	{
		//Open only for reading
		flomod.diskf = fopen( flomod.diskfname, "rb" );
	}

	//If file could not be opened
	if ( flomod.diskf == NULL )
	{
		fprintf( stderr, "%s: could not open '%s'\n", flomod.exename, flomod.diskfname );
		exit( 1 );
	}

	//Get file length
	if ( fseek( flomod.diskf, 0, SEEK_END ) )
	{
		fprintf( stderr, "%s: seek error!\n", flomod.exename );
		exit( 1 );
	}
	flomod.disklen = ftell( flomod.diskf );

	//Seek to place pointed by start descriptor
	if ( fseek( flomod.diskf, flomod.start.offset, SEEK_SET ) )
	{
		fprintf( stderr, "%s: seek error!\n", flomod.exename );
		exit( 1 );
	}

	//Calculate end descriptor
	str2chsb( &flomod.end );
	if ( chsbnull( &flomod.end ) ) //Skip if set by user
	{
		if ( chsbnull( &flomod.length ) )
		{
			//Calculate from detected file size
			flomod.end.offset = flomod.disklen - 1;
			lba2chsb( &flomod.end, &flomod.geom );
			fprintf( stderr, "%s: no endpoint specified - reading whole file\n", flomod.exename );
		}
		else
		{
			//Calculate by adding length and start
			flomod.end = flomod.start;
			chsbsum( &flomod.end, &flomod.length );
		}
	}
	chsbopt( &flomod.end, &flomod.geom );
	chsb2lba( &flomod.end, &flomod.geom );
	chsb2str( &flomod.end );

	//Output start and end position
	if ( flomod.flags & FLOMOD_FLAG_VERBOSE ) fprintf( stderr, "%s: start at %16s - %8ld\n", flomod.exename, flomod.start.str, flomod.start.offset );
	if ( flomod.flags & FLOMOD_FLAG_VERBOSE ) fprintf( stderr, "%s: end at   %16s - %8ld\n", flomod.exename, flomod.end.str, flomod.end.offset );

	long offset = 0;
	if ( flomod.flags & FLOMOD_FLAG_WRITE )
	{
		//Write to disk file
		while ( ( ( b = getchar( ) ) != EOF ) && ( ( offset = ftell( flomod.diskf ) ) <= flomod.end.offset ) )
		{
			if ( flomod.flags & FLOMOD_FLAG_DUMP ) dump( flomod.start.offset, offset, b );
			fputc( b, flomod.diskf );
		}
	}
	else
	{
		//Read from disk file
		while ( ( ( b = fgetc( flomod.diskf ) ) != EOF ) && ( ( offset = ftell( flomod.diskf ) ) <= flomod.end.offset ) )
			if ( flomod.flags & FLOMOD_FLAG_DUMP ) dump( flomod.start.offset, offset - 1, b );
			else putchar( b );
	}

	//Additional newline after dump
	if ( flomod.flags & FLOMOD_FLAG_DUMP ) printf( "\n" );

	//EOF message
	if ( b == EOF && flomod.flags & FLOMOD_FLAG_VERBOSE ) fprintf( stderr, "%s: EOF reached!\n", flomod.exename );

	//Close disk file
	fclose( flomod.diskf );
	return 0;
}


src/flomod.h
/*
	flomod - a floppy disk image editor
	Copyright (C) 2017 Jacek Wieczorek <mrjjot@gmail.com>
	This file is part of flomod.
	Flomod is free software: you can redistribute it and/or modify
	it under the terms of the GNU General Public License as published by
	the Free Software Foundation, either version 3 of the License, or
	(at your option) any later version.
	Flomod is distributed in the hope that it will be useful,
	but WITHOUT ANY WARRANTY; without even the implied warranty of
	MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
	GNU General Public License for more details.
	You should have received a copy of the GNU General Public License
	along with this program.  If not, see <http://www.gnu.org/licenses/>.
*/

#ifndef FLOMOD_H
#define FLOMOD_H
#include <stdio.h>
#include "chsb.h"

#define FLOMOD_FLAG_WRITE 1
#define FLOMOD_FLAG_SECTOR_BASE1 2
#define FLOMOD_FLAG_DEFAULT_FLOPPY 4
#define FLOMOD_FLAG_VERBOSE 8
#define FLOMOD_FLAG_DUMP 16

//Flomod configuration structure
struct flomod
{
	unsigned int flags;
	const char *exename;

	char *disktype;
	char *diskfname;
	FILE *diskf;
	long disklen;
	CHSB start;
	CHSB length;
	CHSB end;
	CHSB geom;
};

#endif


src/disktypes.h
/*
	flomod - a floppy disk image editor
	Copyright (C) 2017 Jacek Wieczorek <mrjjot@gmail.com>
	This file is part of flomod.
	Flomod is free software: you can redistribute it and/or modify
	it under the terms of the GNU General Public License as published by
	the Free Software Foundation, either version 3 of the License, or
	(at your option) any later version.
	Flomod is distributed in the hope that it will be useful,
	but WITHOUT ANY WARRANTY; without even the implied warranty of
	MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
	GNU General Public License for more details.
	You should have received a copy of the GNU General Public License
	along with this program.  If not, see <http://www.gnu.org/licenses/>.
*/

#ifndef DISKTYPES_H
#define DISKTYPES_H
#include "chsb.h"

#define DISKTYPE_FLOPPY_35_740  0
#define DISKTYPE_FLOPPY_35_144  1
#define DISKTYPE_FLOPPY_35_288  2
#define DISKTYPE_FLOPPY_525_12  3
#define DISKTYPE_FLOPPY_525_360 4

const CHSB disktypes[] =
{
	//3.5 inch, 720KB floppy
	[DISKTYPE_FLOPPY_35_740] = {
		.c = 80,
		.h = 2,
		.s = 9,
		.b = 512,
		.offset = 0,
		.str = "FLOPPY_3.5_740K",
		.flags = 0,
	},

	//3.5 inch, 1.44MB floppy
	[DISKTYPE_FLOPPY_35_144] = {
		.c = 80,
		.h = 2,
		.s = 18,
		.b = 512,
		.offset = 0,
		.str = "FLOPPY_3.5_1.44M",
		.flags = 0,
	},

	//3.5 inch, 2.88MB floppy
	[DISKTYPE_FLOPPY_35_288] = {
		.c = 80,
		.h = 2,
		.s = 36,
		.b = 512,
		.offset = 0,
		.str = "FLOPPY_3.5_2.88M",
		.flags = 0,
	},

	//5.25 inch, 1.2MB floppy
	[DISKTYPE_FLOPPY_525_12] = {
		.c = 80,
		.h = 2,
		.s = 15,
		.b = 512,
		.offset = 0,
		.str = "FLOPPY_5.25_1.2M",
		.flags = 0,
	},

	//5.25 inch, 360KB floppy
	[DISKTYPE_FLOPPY_525_360] = {
		.c = 40,
		.h = 2,
		.s = 9,
		.b = 512,
		.offset = 0,
		.str = "FLOPPY_5.25_360K",
		.flags = 0,
	},
};

#define DISKTYPE_COUNT ( (int) ( sizeof( disktypes ) / sizeof( disktypes[0] ) ) )

#endif
